fx.go

```diff
@@ -18,4 +18,150 @@
+	components     []Component
+	scopeCloser    io.Closer
+	logger         *zap.Logger
+	loggerCloserFn func()
+}
+
+// scopeInit is used by the metrics package
+// TODO: Remove after metrics refactor
+type scopeInit struct {
+	config config.Provider
+}
+
+// Name returns the name
+func (c *scopeInit) Name() string {
+	return c.config.Get(config.ServiceNameKey).AsString()
+}
+
+// Config returns the config
+func (c *scopeInit) Config() config.Provider {
+	return c.config
+}
+
+// New creates a service with the provided modules
+func New(modules ...Module) *Service {
+	s := &Service{
+		g:       dig.New(),
+		modules: modules,
+	}
+
+	// load config for module creation and include it in the graph
+	cfg := config.DefaultLoader.Load()
+	// TODO: need to pull latest dig for direct Interface injection fix
+	s.g.MustRegister(func() config.Provider { return cfg })
+
+	scope, _, scopeCloser := metrics.RootScope(&scopeInit{config: cfg})
+	metrics.Freeze()
+	s.g.MustRegister(func() (tally.Scope, error) { return scope, nil })
+	s.scopeCloser = scopeCloser
+
+	// Set up the logger, remember it on the service and also inject into the graph
+	l, closerFn, err := createLogger(cfg, scope)
+	if err != nil {
+		panic("failed to instantiate logger")
+	}
+	s.logger = l
+	s.loggerCloserFn = closerFn
+	s.g.MustRegister(l)
+
+	// add a bunch of stuff
+	for _, m := range modules {
+		for _, ctor := range m.Constructor() {
+			s.g.MustRegister(ctor)
+		}
+	}
+
+	return s
+}
+
+func createLogger(cp config.Provider, scope tally.Scope) (*zap.Logger, func(), error) {
+	cfg := cp.Get("logging")
+	logConfig := ulog.Configuration{}
+	if cfg.HasValue() {
+		if err := logConfig.Configure(cfg); err != nil {
+			return nil, nil, errors.Wrap(err, "failed to initialize logging from config")
+		}
+	} else {
+		// if no config, default to the regular one
+		logConfig = ulog.DefaultConfiguration()
+	}
+
+	logger, err := logConfig.Build(zap.Hooks(ulog.Metrics(scope)))
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to build the logger")
+	}
+	closerFn := ulog.SetLogger(logger)
+	return logger, closerFn, err
+}
+
+// WithComponents adds additional components to the service
+func (s *Service) WithComponents(components ...Component) *Service {
+	s.components = append(s.components, components...)
+	// Add provided components to dig
+	for _, c := range components {
+		s.g.MustRegister(c)
+	}
+	return s
+}
+
+// Start starts the service
+func (s *Service) Start() {
+	// TODO: move to dig, perhaps #Call(constructor) function
+	for _, m := range s.modules {
+		for _, ctor := range m.Constructor() {
+			ctype := reflect.TypeOf(ctor)
+			switch ctype.Kind() {
+			case reflect.Func:
+				objType := ctype.Out(0)
+				s.g.MustResolve(reflect.New(objType).Interface())
+			}
+		}
+		s.logger.Info("Module started", zap.String("module", m.Name()))
+	}
+	s.logger.Info("Service has started")
+}
+
+// Stop stops the service
+func (s *Service) Stop() {
+	for _, m := range s.modules {
```

Comment:
I'd think we need to Stop them in a reversed order, otherwise it can lead to surprises on stop/shutdowns if there are interdependencies between modules.